core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/AmbariDecommissioner.java

```diff
@@ -348,23 +357,22 @@ private void verifyNodeCount(int replication, int scalingAdjustment, List<HostMe
         }
     }
 
-    private Map<String, Long> selectNodes(Map<String, Long> sortedAscending, List<HostMetadata> filteredHostList, int removeCount) {
+    protected Map<String, Long> selectNodes(Map<String, Long> sortedAscending, List<HostMetadata> filteredHostList, int removeCount) {
         LOGGER.info("sortedAscending: {}, filteredHostList: {}", sortedAscending, filteredHostList);
-        Map<String, Long> select = new HashMap<>();
-        int i = 0;
-        for (Entry<String, Long> entry : sortedAscending.entrySet()) {
-            if (i < removeCount) {
-                for (HostMetadata hostMetadata : filteredHostList) {
-                    if (hostMetadata.getHostName().equalsIgnoreCase(entry.getKey())) {
-                        select.put(entry.getKey(), entry.getValue());
-                        i++;
-                        break;
-                    }
-                }
-            } else {
-                break;
-            }
-        }
+
+        Map<String, Long> select  = filteredHostList
+                .stream()
+                .filter(hostMetadata -> hostMetadata.getHostMetadataState() == HostMetadataState.UNHEALTHY
+                        && sortedAscending.containsKey(hostMetadata.getHostName()))
+                .limit(removeCount)
+                .collect(Collectors.toMap(HostMetadata::getHostName, o -> 0L));
+
+        sortedAscending.entrySet().stream()
+                .filter(a -> select.size() < removeCount)
```

Comment:
Filter will happen first, that means the result of `select.size() < removeCount` is constant. Later on `forEach()` changes `select` but it runs on the filtered stream.